modules/tensor_mechanics/src/materials/ADIsotropicPlasticityStressUpdate.C

```diff
@@ -0,0 +1,238 @@
+
+template <ComputeStage compute_stage>
+void
+ADIsotropicPlasticityStressUpdate<compute_stage>::computeStressInitialize(
+    const ADReal & effective_trial_stress, const ADRankFourTensor & elasticity_tensor)
+{
+  computeYieldStress(elasticity_tensor);
+
+  _yield_condition = effective_trial_stress - _hardening_variable_old[_qp] - _yield_stress;
+  _hardening_variable[_qp] = _hardening_variable_old[_qp];
+  _plastic_strain[_qp] = _plastic_strain_old[_qp];
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADIsotropicPlasticityStressUpdate<compute_stage>::computeResidual(
+    const ADReal & effective_trial_stress, const ADReal & scalar)
+{
+  ADReal residual = 0.0;
+
+  mooseAssert(_yield_condition != -1.0,
+              "the yield stress was not updated by computeStressInitialize");
+
+  if (_yield_condition > 0.0)
+  {
+    _hardening_slope = computeHardeningDerivative(scalar);
+    _hardening_variable[_qp] = computeHardeningValue(scalar);
+
+    residual =
+        (effective_trial_stress - _hardening_variable[_qp] - _yield_stress) / _three_shear_modulus -
+        scalar;
+  }
+  return residual;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADIsotropicPlasticityStressUpdate<compute_stage>::computeDerivative(
+    const ADReal & /*effective_trial_stress*/, const ADReal & /*scalar*/)
+{
+  if (_yield_condition > 0.0)
+    return -1.0 - _hardening_slope / _three_shear_modulus;
+
+  return 1.0;
+}
+
+template <ComputeStage compute_stage>
+void
+ADIsotropicPlasticityStressUpdate<compute_stage>::iterationFinalize(ADReal scalar)
+{
+  if (_yield_condition > 0.0)
+    _hardening_variable[_qp] = computeHardeningValue(scalar);
+}
+
+template <ComputeStage compute_stage>
+void
+ADIsotropicPlasticityStressUpdate<compute_stage>::computeStressFinalize(
+    const ADRankTwoTensor & plasticStrainIncrement)
+{
+  _plastic_strain[_qp] += plasticStrainIncrement;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADIsotropicPlasticityStressUpdate<compute_stage>::computeHardeningValue(const ADReal & scalar)
+{
+  if (_hardening_function)
+  {
+    const Real strain_old = _effective_inelastic_strain_old[_qp];
+    Point p;
+
+    Real t = strain_old + MetaPhysicL::raw_value(scalar);
+    return _hardening_function->value(t, p) - _yield_stress;
+  }
+
+  return _hardening_variable_old[_qp] + (_hardening_slope * scalar);
+}
+
+template <>
+DualReal
+ADIsotropicPlasticityStressUpdate<JACOBIAN>::computeHardeningValue(const DualReal & scalar)
+{
+  if (_hardening_function)
+  {
+    const Real strain_old = _effective_inelastic_strain_old[_qp];
+    Point p;
+
+    Real t = strain_old + MetaPhysicL::raw_value(scalar);
+    DualReal hardening_function_value = _hardening_function->value(t, p);
+    hardening_function_value.derivatives() =
+        (strain_old + scalar).derivatives() * _hardening_function->timeDerivative(t, p);
+    return hardening_function_value - _yield_stress;
+  }
+
+  return _hardening_variable_old[_qp] + (_hardening_slope * scalar);
```

Comment:
extra ()? Probably not worth changing since it's probably just copy-pasted from the the non-AD. Just noting it.

Reply:
Yes, almost all of those comments are regarding copy-pasted code. I'm changing both files. The comments are all on point and valuable.

Reply:
Yeah, I remember it helped to have them exactly the same so you could diff them.